### PR TITLE
C#: Add query for missing function level access control 

### PR DESCRIPTION
**Pull request by joefarebrother**
Adds query for missing authorization checks.
Finds WebForms methods (`System.Web.UI.Page`) and MVC actions whose names indicate that they should have authorization checs (e.g. `Delete` or `Admin`), but such checks could not be found; including through code, attributes, or `Web.config` xml files.

### CODE REVIEW

File: csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  )
+}
+
+/** Holds if the filepath `route` can refer to `actual` after expanding a '~". */
+bindingset[route, actual]
+private predicate physicalRouteMatches(string route, string actual) {
+  route = actual
+  or
+  route.charAt(0) = "~" and
+  exists(string dir | actual = dir + route.substring(1, route.length()) + ".cs")
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:96]:
```suggestion
  actual.matches("%" + route.suffix(1) + ".cs")
```

File: csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    or
+    this.(MethodCall).getTarget().getName().toLowerCase().matches("%auth%")
+    or
+    this.(PropertyAccess).getTarget().getName().toLowerCase().matches("%auth%")
+  }
+}
+
+/** Holds if `m` is a method that should have an auth check, and does indeed have one. */
+predicate hasAuthViaCode(ActionMethod m) {
+  m.needsAuth() and
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:118]: Why is this needed here (`missingAuth` also checks this)?

File: csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  exists(Callable caller, AuthExpr auth |
+    m.getAnAuthorizingCallable().calls*(caller) and
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:120]: Could this cause performance issues (the reflexive and transitive closure of this could be huge)?

File: csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  }
+}
+
+/**
+ * Holds if the given action has an xml `authorization` tag that refers to it.
+ */
+predicate hasAuthViaXml(ActionMethod m) {
+  exists(AuthorizationXmlElement el, string rest |
+    el.hasDenyElement() and
+    m.getARoute() = el.getARoute() + rest
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:165]:
```suggestion
  exists(AuthorizationXmlElement el |
    el.hasDenyElement() and
    m.getARoute().matches(el.getARoute() + "%")
  )```

File: csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    attr = m.getDeclaringType().getABaseType*().getAnAttribute()
+  )
+}
+
+/** Holds if `m` is a method that should have an auth check, but is missing it. */
+predicate missingAuth(ActionMethod m) {
+  m.needsAuth() and
+  not hasAuthViaCode(m) and
+  not hasAuthViaXml(m) and
+  not hasAuthViaAttribute(m) and
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:182]: Maybe add some test cases for this part as well.

**Review by michaelnebel — COMMENTED** [review]
This already looks really good!
I have added some minor comments/recommendations and suggestions.
There primary concern is probably about the use of `.calls*(..)`.
Have you tried running the query on some large projects?

**Comment by michaelnebel** [comment]
> This already looks really good! I have added some minor comments/recommendations and suggestions. There primary concern is probably about the use of `.calls*(..)`. Have you tried running the query on some large projects?

It is probably a good idea to run this query against the DCA nightly suite.

File: csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll (hunk 1 above)
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:95]: Would that not work if the string had a % in it?

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:95]: Auch - you are right. If route contains any of the regex *special* characters then this doesn't work!

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:95]: Maybe just
```suggestion
  exists(string dir | actual = dir + route.suffix(1) + ".cs")
```

File: csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll (hunk 3 above)
Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:119]: Alright - that sounds good!

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:119]: @hvitved: Do you have any concerns about using the reflexive, transitive closure of `calls`?

Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll:119]: If DCA is happy, then I'm happy :-)

**Review by michaelnebel — APPROVED** [review]
Great work Joe!!! :-)
Looks good to me!

**Comment by joefarebrother** [comment]
Great; just need the docs review then

File: csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+Sensitive actions, such as editing or deleting content, or accessing admin pages, should have authentication checks 
+to ensure that they cannot be used by arbitrary users.
+</p>
+
+</overview>
+<recommendation>
+
+<p>
+Ensure that proper authorization checks are made for sensitive actions. 
+For WebForms applications, the <code>authorazation</code> tag in <code>Web.config</code> XML files
```
Comment (tamasvajk) [inline csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp:17]: There's a typo in `authorazation`.

File: csharp/ql/src/Security Features/CWE-285/MVC.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+public class ProfileController : Controller {
+
+    // BAD: No authorization is used.
+    public ActionResult Edit(int id) {
+        ...
+    }
+
+    // GOOD: The `Authorize` tag is used.
```
Comment (tamasvajk) [inline csharp/ql/src/Security Features/CWE-285/MVC.cs:8]: Minor: tag should be changed to attribute.

**Review by tamasvajk — COMMENTED** [review]
Looks good, I found two small issues.

**Comment by mchammer01** [comment]
I'll review this for Docs!

File: csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+</p>
+
+</overview>
+<recommendation>
+
+<p>
+Ensure that proper authorization checks are made for sensitive actions. 
+For WebForms applications, the <code>authorazation</code> tag in <code>Web.config</code> XML files 
+can be used to implement access control. The <code>System.Web.UI.Page.User</code> property can also be 
+used to verify a user's roles.
```
Comment (mchammer01) [inline csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp:19]: Can a user have several roles? 🤔 
```suggestion
used to verify a user's role. 
```

File: csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+<p>
+In the following MVC example, the case marked BAD has no authorization
+checks; whereas the case marked GOOD uses the <code>Authorize</code> attribute.
+</p>
+
+<sample src="MVC.cs" />
+
+</example>
+<references>
+<li><code>Page.User</code> Property - <a href="https://learn.microsoft.com/en-us/dotnet/api/system.web.ui.page.user?view=netframework-4.8.1#system-web-ui-page-user">Microsoft Learn</a></li>
```
Comment (mchammer01) [inline csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp:50]: Missing punctuation in list items
```suggestion
<li><code>Page.User</code> Property - <a href="https://learn.microsoft.com/en-us/dotnet/api/system.web.ui.page.user?view=netframework-4.8.1#system-web-ui-page-user">Microsoft Learn</a>.</li>
```

File: csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+used to verify a user's roles. 
+For MVC applications, the <code>Authorize</code> attribute can be used to require authorization on specific 
+action methods.
+</p>
+
+</recommendation>
+<example>
+
+<p>
+In the following WebForms example, the case marked BAD has no authorization checks; whereas the
```
Comment (mchammer01) [inline csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp:28]: No need for `;` as we have `whereas`.
```suggestion
In the following WebForms example, the case marked BAD has no authorization checks whereas the 
```

File: csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+<p>
+The following <code>Web.config</code> file uses the <code>authorization</code> tag to deny access to anonymous users,
+in a <code>location</code>> tag to have it apply to a specific path.
+</p>
+
+<sample src="Web.config" />
+
+<p>
+In the following MVC example, the case marked BAD has no authorization
+checks; whereas the case marked GOOD uses the <code>Authorize</code> attribute.
```
Comment (mchammer01) [inline csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp:43]: Nit (same comment as above)
```suggestion
checks whereas the case marked GOOD uses the <code>Authorize</code> attribute.
```

File: csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+</p>
+
+<sample src="MVC.cs" />
+
+</example>
+<references>
+<li><code>Page.User</code> Property - <a href="https://learn.microsoft.com/en-us/dotnet/api/system.web.ui.page.user?view=netframework-4.8.1#system-web-ui-page-user">Microsoft Learn</a></li>
+<li>Control authorization permissions in an ASP.NET application - <a href="https://learn.microsoft.com/en-us/troubleshoot/developer/webapps/aspnet/www-authentication-authorization/authorization-permissions">Microsoft Learn</a></li>
```
Comment (mchammer01) [inline csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp:51]:
```suggestion
<li>Control authorization permissions in an ASP.NET application - <a href="https://learn.microsoft.com/en-us/troubleshoot/developer/webapps/aspnet/www-authentication-authorization/authorization-permissions">Microsoft Learn</a>.</li>
```

File: csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+<li>Simple authorization in ASP.NET Core - <a href="https://learn.microsoft.com/en-us/aspnet/core/security/authorization/simple?view=aspnetcore-7.0">Microsoft Learn</a></li>
```
Comment (mchammer01) [inline csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp:52]:
```suggestion
<li>Simple authorization in ASP.NET Core - <a href="https://learn.microsoft.com/en-us/aspnet/core/security/authorization/simple?view=aspnetcore-7.0">Microsoft Learn</a>.</li>
```

File: csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+<p>
+In the following WebForms example, the case marked BAD has no authorization checks; whereas the 
+case marked GOOD uses <code>User.IsInRole</code> to check for the user's role.
+</p>
+
+<sample src="WebForms.cs" />
+
+<p>
+The following <code>Web.config</code> file uses the <code>authorization</code> tag to deny access to anonymous users,
+in a <code>location</code>> tag to have it apply to a specific path.
```
Comment (mchammer01) [inline csharp/ql/src/Security Features/CWE-285/MissingAccessControl.qhelp:36]: Typo + would it be feasible to update the language here. What does "it" refer to?
```suggestion
in a <code>location</code> tag to have it apply to a specific path.
```

**Review by mchammer01 — COMMENTED** [review]
@joefarebrother 👋🏻 - I wasn't able to see a preview of the ql file, with the examples in-situ. Have things changed, do you know how to get to the preview?
I reviewed this from an editorial point of view, and left a few comments and suggestions for your consideration. Feel free to ignore the ones you don't agree with 😃 
(The punctuation is needed in the list in the References section. )

**Review by mchammer01 — APPROVED** [review]
LGTM :shipit:
